src/x509.lisp

```diff
@@ -0,0 +1,358 @@
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defconstant +V-ASN1-OCTET-STRING+ 4)
+  (defconstant +V-ASN1-UTF8STRING+ 12)
+  (defconstant +V-ASN1-PRINTABLESTRING+ 19)
+  (defconstant +V-ASN1-TELETEXSTRING+ 20)
+  (defconstant +V-ASN1-IASTRING+ 22)
+  (defconstant +V-ASN1-UNIVERSALSTRING+ 28)
+  (defconstant +V-ASN1-BMPSTRING+ 30))
+
+
+(defconstant +NID-subject-alt-name+ 85)
+(defconstant +NID-commonName+   13)
+
+(cffi:defcstruct general-name
+  (type :int)
+  (data :pointer))
+
+(cffi:defcfun ("sk_value" sk-value)
+    :pointer
+  (stack :pointer)
+  (index :int))
+
+(cffi:defcfun ("sk_num" sk-num)
+    :int
+  (stack :pointer))
+
+(declaim (ftype (function (cffi:foreign-pointer fixnum) cffi:foreign-pointer) sk-general-name-value))
+(defun sk-general-name-value (names index)
+  (sk-value names index))
+
+(declaim (ftype (function (cffi:foreign-pointer) fixnum) sk-general-name-num))
+(defun sk-general-name-num (names)
+  (sk-num names))
+
+(cffi:defcfun ("GENERAL_NAMES_free" general-names-free)
+    :void
+  (general-names :pointer))
+
+(cffi:defcfun ("ASN1_STRING_data" asn1-string-data)
+    :pointer
+  (asn1-string :pointer))
+
+(cffi:defcfun ("ASN1_STRING_length" asn1-string-length)
+    :int
+  (asn1-string :pointer))
+
+(cffi:defcfun ("ASN1_STRING_type" asn1-string-type)
+    :int
+  (asn1-string :pointer))
+
+(cffi:defcstruct asn1_string_st
+  (length :int)
+  (type :int)
+  (data :pointer)
+  (flags :long))
+
+#|
+ASN1 string validation references:
+ - https://github.com/digitalbazaar/forge/blob/909e312878838f46ba6d70e90264650b05eb8bde/js/asn1.js
+ - http://www.obj-sys.com/asn1tutorial/node128.html
+ - https://github.com/deadtrickster/ssl_verify_hostname.erl/blob/master/src/ssl_verify_hostname.erl
+ - https://golang.org/src/encoding/asn1/asn1.go?m=text
+|#
+(defgeneric decode-asn1-string (asn1-string type))
+
+(defun copy-bytes-to-lisp-vector (src-ptr vector count)
+  (declare (type (simple-array (unsigned-byte 8)))
+           (type fixnum count)
+           (optimize (safety 0) (debug 0) (speed 3)))
+  (dotimes (i count vector)
+    (setf (aref vector i) (cffi:mem-aref src-ptr :unsigned-char i))))
+
+(defun asn1-string-bytes-vector (asn1-string)
+  (let* ((data (asn1-string-data asn1-string))
+         (length (asn1-string-length asn1-string))
+         (vector (cffi:make-shareable-byte-vector length)))
+    (copy-bytes-to-lisp-vector data vector length)
+    vector))
+
+(declaim (inline asn1-iastring-char-p))
+(defun asn1-iastring-char-p (byte)
+  (declare (type (unsigned-byte 8) byte)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (< byte #x80))
+
+(defun asn1-iastring-p (bytes)
+  (declare (type (simple-array (unsigned-byte 8)) bytes)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (every #'asn1-iastring-char-p bytes))
+
+(defmethod decode-asn1-string (asn1-string (type (eql #.+v-asn1-iastring+)))
```

Comment:
I remember there were some problems on lisps other than sbcl, probably clisp. that is why respective defconstants wrapped in eval-when, but there is a possibility that #. was used before eval-when introduced and therefore obsolete now